src/test/solution_space/solution_graph/graph_vertices_dist_tests.py

```diff
@@ -0,0 +1,186 @@
+
+
+# Finds distance matrix between all vertices in given dict, using anon trees which indices listed in the dict.
+#
+# For example, for added_indices_by_vertex: {VERTEX.VERTEX_0: [0, 3],
+#                                            VERTEX.VERTEX_1: [4]}
+# it will find all distances between
+# 1. VERTEX_0 with fragment_0, fragment_3
+# 2. VERTEX_1 with fragment_4.
+#
+# It's useful for finding intermediate distance matrix while adding new fragments.
+def find_dist_matrix(added_indices_by_vertex: Dict[VERTEX, List[int]]) -> List[List[int]]:
+    dist_matrix = []
+    for i, src_vertex in enumerate(added_indices_by_vertex.keys()):
+        dist_matrix.append([])
+        src_indices = added_indices_by_vertex[src_vertex]
+        for j, dst_vertex in enumerate(added_indices_by_vertex.keys()):
+            dst_indices = added_indices_by_vertex[dst_vertex]
+            canon_dist = canon_distance[src_vertex][dst_vertex]
+            anon_dist = min([anon_distance[k][l] for k in src_indices for l in dst_indices])
+            dist_matrix[i].append(min(anon_dist, canon_dist))
+    return dist_matrix
+
+
+def get_code_info_chain(sources: List[str]) -> List[Tuple[Code, CodeInfo]]:
+    user = User()
+    return [(create_code_from_source(s), CodeInfo(user)) for s in sources]
+
+
+@pytest.mark.skipif(to_skip(current_module_level=TEST_LEVEL.SOLUTION_SPACE), reason=TEST_LEVEL.SOLUTION_SPACE.value)
+class TestDistBetweenVertices():
+
+    # Check all fragments in same vertices have the same canon trees
+    @pytest.mark.parametrize('vertex', [v for v in VERTEX])
+    def test_same_canon_trees_in_same_vertices(self, vertex: VERTEX):
+        same_canon_fragments = [all_fragments[i] for i in INDICES_BY_VERTEX[vertex]]
+        canon_trees = [get_trees(f, {TREE_TYPE.CANON})[0] for f in same_canon_fragments]
+        for canon_tree_1, canon_tree_2 in itertools.product(canon_trees, repeat=2):
+            self.assertTrue(are_asts_equal(canon_tree_1, canon_tree_2))
+
+    # Check all fragments in different vertices have different canon trees
+    def test_different_canon_trees_in_different_vertices(self):
+        # Take the first fragment from each vertex to get all fragments with different canon trees
+        different_canon_fragments = [all_fragments[INDICES_BY_VERTEX[vertex][0]] for vertex in VERTEX]
+        canon_trees = [get_trees(f, {TREE_TYPE.CANON})[0] for f in different_canon_fragments]
+        for canon_tree_1, canon_tree_2 in zip(canon_trees, np.roll(canon_trees, 1)):
+            self.assertFalse(are_asts_equal(canon_tree_1, canon_tree_2))
+
+    # Check anon distance matrix is filled right
+    def test_anon_distance_correctness(self):
+        fragments = [fragment_0, fragment_1, fragment_2, fragment_3, fragment_4, fragment_5]
+        for i, src_fragment in enumerate(fragments):
+            src_anon_tree, = get_trees(src_fragment, {TREE_TYPE.ANON})
+            for j, dst_fragment in enumerate(fragments):
```

Comment:
Please use [subtests](https://github.com/pytest-dev/pytest-subtests) for separating tests. You can also see an example in the `diff_handler_test.py`